### ISSUE TITLE
If a client resets the TCP connection after receiving an HTTP response, an Akka server leaks one ActorInterpreter

### ISSUE THREAD

**Issue by catalin-ursachi** [akka/akka#17551]
This problem has appeared after we upgraded our app from Akka Streams M4 (forked to include https://github.com/akka/akka/pull/17035) to a snapshot of release-2.3-dev at commit c0474d847b (from 20 May).

We are using a custom-made client to test our app. This is an example wire trace of a request made by this client.

```
No.     Time                       Source                Destination           Protocol Length Info
    671 2015-05-21 20:36:10.284759 client                server                TCP      74     40095→8823 [SYN] Seq=0 Win=14600 Len=0 MSS=1460 SACK_PERM=1 TSval=1651557437 TSecr=0 WS=128
    672 2015-05-21 20:36:10.284803 server                client                TCP      74     8823→40095 [SYN, ACK] Seq=0 Ack=1 Win=14480 Len=0 MSS=1460 SACK_PERM=1 TSval=3717873627 TSecr=1651557437 WS=128
    673 2015-05-21 20:36:10.285105 client                server                TCP      66     40095→8823 [ACK] Seq=1 Ack=1 Win=14720 Len=0 TSval=1651557438 TSecr=3717873627
    674 2015-05-21 20:36:10.286598 client                server                HTTP     683    GET http://server.hostname/requestPath HTTP/1.1 
    675 2015-05-21 20:36:10.286623 server                client                TCP      66     8823→40095 [ACK] Seq=1 Ack=618 Win=15744 Len=0 TSval=3717873629 TSecr=1651557439
    803 2015-05-21 20:36:10.340009 server                client                TCP      201    [TCP segment of a reassembled PDU]
    804 2015-05-21 20:36:10.340059 server                client                HTTP     962    HTTP/1.1 200 OK  (application/json)
    805 2015-05-21 20:36:10.340373 client                server                TCP      66     40095→8823 [ACK] Seq=618 Ack=136 Win=15744 Len=0 TSval=1651557493 TSecr=3717873683
    806 2015-05-21 20:36:10.340415 client                server                TCP      66     40095→8823 [ACK] Seq=618 Ack=1032 Win=17536 Len=0 TSval=1651557493 TSecr=3717873683
    808 2015-05-21 20:36:10.342456 client                server                TCP      66     40095→8823 [RST, ACK] Seq=618 Ack=1032 Win=17536 Len=0 TSval=1651557495 TSecr=3717873683
```

This causes the server to leak one ActorInterpreter.

Note that, after receiving the response, the client sends two ACKs and an (RST, ACK). For comparison, the wire trace of an otherwise identical request made using an Akka-based client is:

```
No.     Time                       Source                Destination           Protocol Length Info
  35031 2015-05-21 23:00:22.897091 client                server                TCP      66     64818→8823 [SYN] Seq=0 Win=8192 Len=0 MSS=1460 WS=256 SACK_PERM=1
  35032 2015-05-21 23:00:22.897147 server                client                TCP      66     8823→64818 [SYN, ACK] Seq=0 Ack=1 Win=14600 Len=0 MSS=1460 SACK_PERM=1 WS=128
  35033 2015-05-21 23:00:22.897640 client                server                TCP      60     64818→8823 [ACK] Seq=1 Ack=1 Win=65536 Len=0
  35036 2015-05-21 23:00:22.900206 client                server                HTTP     749    GET /requestPath HTTP/1.1 
  35037 2015-05-21 23:00:22.900257 server                client                TCP      54     8823→64818 [ACK] Seq=1 Ack=696 Win=16000 Len=0
  35684 2015-05-21 23:00:23.036501 server                client                TCP      190    [TCP segment of a reassembled PDU]
  35686 2015-05-21 23:00:23.036644 server                client                HTTP     1313   HTTP/1.1 200 OK  (application/json)
  35690 2015-05-21 23:00:23.037615 client                server                TCP      60     64818→8823 [ACK] Seq=696 Ack=1396 Win=64256 Len=0
  35743 2015-05-21 23:00:23.046091 client                server                TCP      60     64818→8823 [FIN, ACK] Seq=696 Ack=1396 Win=64256 Len=0
  35750 2015-05-21 23:00:23.047216 server                client                TCP      54     8823→64818 [FIN, ACK] Seq=1396 Ack=697 Win=16000 Len=0
  35754 2015-05-21 23:00:23.047873 client                server                TCP      60     64818→8823 [ACK] Seq=697 Ack=1397 Win=64256 Len=0
```

This doesn't result in any ActorInterpreter leak.

**Comment by catalin-ursachi** [comment]
The following test reproduces the bug - i.e., it passes under versions where the bug doesn't exist, and fails under the latest version of Akka:

```scala
package akka.http

import akka.actor._
import akka.http.scaladsl.Http
import akka.http.scaladsl.client.RequestBuilding
import akka.http.scaladsl.model._
import akka.pattern.after
import akka.stream.ActorFlowMaterializer
import akka.stream.impl.fusing.ActorInterpreter
import akka.stream.scaladsl.{Flow, Sink, Source}
import akka.util.ByteString
import org.specs2.mutable.Specification
import org.specs2.time.NoTimeConversions
import scala.collection.mutable
import scala.concurrent.duration._
import scala.concurrent.{Await, Future}
import scala.util.Success

class ActorLeakOnConnectionFailureSpec
  extends Specification
  with RequestBuilding
  with NoTimeConversions {

  val defaultTimeout = 20.seconds

  /**
   * A simple Akka server, which immediately responds with 200 OK, but sends the response body only after a delay.
   */
  class TestServer() {

    implicit val actorSystem = ActorSystem("ServerActorSystem")
    implicit val materializer = ActorFlowMaterializer(namePrefix = Some("ServerMaterializer"))

    val responseBody = ByteString("Response body")
    val responseLatency = 3.seconds

    def slowSource() = Source(after(responseLatency, actorSystem.scheduler)(Future.successful(responseBody)))

    val testServerFuture = Http().bind("127.0.0.1", 0).to {
      Sink.foreach { incomingConnection =>
        incomingConnection.handleWith(Flow[HttpRequest].map { request =>
          HttpResponse(entity = HttpEntity(ContentTypes.`application/octet-stream`, responseBody.size, slowSource()))
        })
      }
    }.run()

    val server = Await.result(testServerFuture, defaultTimeout)

    val host: String = server.localAddress.getHostName
    val port: Int = server.localAddress.getPort
  }

  /**
   * A simple test client which sends requests to a particular host and port.
   */
  class TestClient(host: String, port: Int) {

    implicit val actorSystem = ActorSystem("ClientActorSystem")
    implicit val materializer = ActorFlowMaterializer(namePrefix = Some("clientMaterializer"))

    def sendRequest(request: HttpRequest): Future[HttpResponse] = {

      val connection =
        Http().outgoingConnection(host, port)

      Source.single(request) via connection runWith Sink.head
    }
  }

  "A simple Akka server" should {

    "not leak ActorInterpreters if the client resets the connection" in {
      // Given
      implicit val testSystem = ActorSystem("TestSystem")
      implicit val materializer = ActorFlowMaterializer()

      val server = new TestServer()
      val client = new TestClient(server.host, server.port)

      val request = Get(s"http://${server.host}:${server.port}")

      // We make an initial request, and stream in the response. This is to ensure all flows are initialized.
      client.sendRequest(request)
        .flatMap(_.entity.toStrict(defaultTimeout))
        .map(_.data) must
        beEqualTo(server.responseBody).await(timeout = defaultTimeout)

      val initialActorCount: Int = countActorsInSystem(server.actorSystem)

      // We make a request, and shut down the client's actor system immediately after receiving the metadata -
      // that is to say, before the server has had time to dispatch the response body.
      //
      // This will cause the server to leak one [[ActorInterpreter]]
      client.sendRequest(request)
        .andThen { case Success(response) =>
        client.actorSystem.shutdown()
        Thread.sleep(server.responseLatency.toMillis * 2)
      } must be.await(timeout = defaultTimeout)

      // Then
      /**
       * We wait to allow all streams enough time to complete. Under a version where the bug doesn't exist,
       * 1 second is enough for the [[ActorInterpreter]] count to fall back to its initial value.
       */
      val cooldownAllowance = 1.seconds
      Thread.sleep(cooldownAllowance.toMillis)

      val finalActorCount = countActorsInSystem(server.actorSystem)

      /**
       * In the latest version, this will fail - the final count will be precisely 1 greater than the initial count.
       */
      finalActorCount must not beGreaterThan initialActorCount

      server.actorSystem.shutdown()
      testSystem.shutdown()
      ok
    }
  }

  /**
   * A method that counts the actors inside a given actor system.
   */
  def countActorsInSystem(system: ActorSystem) = {

    val actorCountsByClass = new mutable.Map.WithDefault[Class[_ <: Actor], Int](mutable.Map.empty, _ => 0)

    def walk(ref: ActorRef): Unit = {
      ref match {
        case c: ActorRefWithCell =>
          c.underlying match {
            case ac: ActorCell =>
              if (ac.actor != null) {
                val actorClass: Class[_ <: Actor] = ac.actor.getClass
                actorCountsByClass.put(actorClass, actorCountsByClass(actorClass) + 1)
              }
              c.children.foreach(walk)

            case unrecognizedCell: Cell =>
              unrecognizedCell.childrenRefs.children.foreach(walk)
          }
        case _ =>
      }
    }

    walk(system.asInstanceOf[ActorSystemImpl].lookupRoot)

    actorCountsByClass.foldLeft(0)(_ + _._2)
  }
}
```

This is the wire trace of the second request:

```
No.     Time                       Source                Destination           Protocol Length Info
    449 2015-05-22 01:32:39.689493 127.0.0.1             127.0.0.1             TCP      52     49697→49693 [SYN] Seq=0 Win=8192 Len=0 MSS=65495 WS=256 SACK_PERM=1
    450 2015-05-22 01:32:39.689493 127.0.0.1             127.0.0.1             TCP      52     49693→49697 [SYN, ACK] Seq=0 Ack=1 Win=8192 Len=0 MSS=65495 WS=256 SACK_PERM=1
    451 2015-05-22 01:32:39.689493 127.0.0.1             127.0.0.1             TCP      40     49697→49693 [ACK] Seq=1 Ack=1 Win=65536 Len=0
    460 2015-05-22 01:32:39.691505 127.0.0.1             127.0.0.1             HTTP     132    GET http://127.0.0.1:49693 HTTP/1.1 
    461 2015-05-22 01:32:39.691505 127.0.0.1             127.0.0.1             TCP      40     49693→49697 [ACK] Seq=1 Ack=93 Win=65536 Len=0
    466 2015-05-22 01:32:39.707504 127.0.0.1             127.0.0.1             TCP      182    [TCP segment of a reassembled PDU]
    467 2015-05-22 01:32:39.707504 127.0.0.1             127.0.0.1             TCP      40     49697→49693 [ACK] Seq=93 Ack=143 Win=65536 Len=0
    472 2015-05-22 01:32:39.712506 127.0.0.1             127.0.0.1             TCP      40     49697→49693 [RST, ACK] Seq=93 Ack=143 Win=0 Len=0
```

**Comment by jrudolph** [comment]
Thanks @catalin-ursachi for the report.

Does your snapshot include the fix for your previous issue from #17529, i.e. commit fe6111b is included? The commit id (c0474d847b) you gave above doesn't seem to be from this repository.

**Comment by drewhk** [comment]
Is this Windows?

**Comment by drewhk** [comment]
If it is Windows, then you can try the the setting "akka.io.windows-connection-abort-workaround-enabled = auto". We turned it off in one of the 2.3.x releases because it introduces performance problems. Since we updated the Akka dependency for streams, this turning off has a potential effect.

**Comment by jrudolph** [comment]
I did only a brief conversion of the test example but so far it seems to work for me on current release-2.3-dev.

**Comment by drewhk** [comment]
This is why I suspect the Windows related issue. See one of my recent PRs: https://github.com/akka/akka/pull/17544

**Comment by catalin-ursachi** [comment]
Hmm. You're right; I'm not sure where I got c0474d847b from; I meant to say 0af8c4205e, which is precisely the merge of https://github.com/akka/akka/pull/17529 to release-2.3-dev.

Otherwise, no: on the version given above, the test fails for me on both Windows and Linux. And the original bug - the actor leak in our actual Akka app - is happening on Linux.

**Comment by drewhk** [comment]
please check at the connection-abort-workaround setting on Windows at least.

**Comment by catalin-ursachi** [comment]
Will do.

One more thing: I'm using Akka branched off at 0af8c42, but set to use Akka version 2.3.11 instead of 2.3.10, as per: https://github.com/akka/akka/issues/17356. (That issue thread suggests that release-2.3-dev would already be using 2.3.11, but rather oddly it isn't).

**Comment by drewhk** [comment]
> That issue thread suggests that release-2.3-dev would already be using 2.3.11, but rather oddly it isn't

https://github.com/akka/akka/pull/17542

**Comment by catalin-ursachi** [comment]
Ah, I see. Thanks.

**Comment by catalin-ursachi** [comment]
I've tried reverting to Akka version 2.3.10. The test still fails.

Is there any chance you could post the modified test that works natively in the Akka project, so I can test it on a couple of versions? :)

**Comment by drewhk** [comment]
I don't think it is related to Akka versions. There are two issues that I can think of:
 - not properly handling connection aborts. On Windows there is a known issue for which I posted the flag which turns on a workaround. 
 - WS actors are still leaking, but the scenario is when the http stream is failed instead of just completed (TCP aborts fail the TCP stream)

**Comment by catalin-ursachi** [comment]
But do either of those explain why the test should fail for me - even with Akka v 2.3.10 and with the 'akka.io.windows-connection-abort-workaround-enabled = auto' flag set - but it passes for @jrudolph on the latest release-2.3-dev?

**Comment by drewhk** [comment]
No, I have no explanation for that. If the connection workaround has no effects, then I don't think it relates to a missed TCP abort. It would be important to know exactly what stage is encapsulated in the leaked InterpreterActor.

**Comment by jrudolph** [comment]
@catalin-ursachi, I guess I broke something in the conversion, so it would be nice if you could check if my code still does something comparable to your original code:

```scala
package akka.http

import akka.http.scaladsl.Http

import scala.concurrent.duration._

import akka.actor.ActorSystem
import akka.http.scaladsl.client.RequestBuilding
import akka.http.scaladsl.model.{ ContentTypes, HttpEntity, HttpResponse, HttpRequest }
import akka.stream.ActorFlowMaterializer
import akka.stream.scaladsl.{ Flow, Sink, Source }
import akka.util.ByteString
import org.scalatest.{ Matchers, WordSpec }

import scala.concurrent.{ Future, Await, Promise }
import scala.util.Success

class IncomingConnectionActorInterpreterLeakSpec2 extends WordSpec with Matchers {

  val defaultTimeout = 20.seconds

  /**
   * A simple Akka server, which immediately responds with 200 OK, but sends the response body only after a delay.
   */
  class TestServer() {

    implicit val actorSystem = ActorSystem("ServerActorSystem")
    implicit val materializer = ActorFlowMaterializer(namePrefix = Some("ServerMaterializer"))

    val responseBody = ByteString("Response body")
    val responseLatency = 3.seconds

    import actorSystem.dispatcher

    val p = Promise[ByteString]()
    actorSystem.scheduler.scheduleOnce(responseLatency) {
      p.complete(Success(responseBody))
    }

    def slowSource() = Source(p.future)

    val testServerFuture = Http().bind("127.0.0.1", 0).to {
      Sink.foreach { incomingConnection ⇒
        incomingConnection.handleWith(Flow[HttpRequest].map { request ⇒
          HttpResponse(entity = HttpEntity(ContentTypes.`application/octet-stream`, responseBody.size, slowSource()))
        })
      }
    }.run()

    val server = Await.result(testServerFuture, defaultTimeout)

    val host: String = server.localAddress.getHostName
    val port: Int = server.localAddress.getPort
  }

  /**
   * A simple test client which sends requests to a particular host and port.
   */
  class TestClient(host: String, port: Int) {

    implicit val actorSystem = ActorSystem("ClientActorSystem")
    implicit val materializer = ActorFlowMaterializer(namePrefix = Some("clientMaterializer"))

    def sendRequest(request: HttpRequest): Future[HttpResponse] = {

      val connection =
        Http().outgoingConnection(host, port)

      Source.single(request) via connection runWith Sink.head
    }
  }

  "A simple Akka server" should {

    "not leak ActorInterpreters after incoming connections are closed" in {
      // Given
      implicit val testSystem = ActorSystem("TestSystem")
      import testSystem.dispatcher
      implicit val materializer = ActorFlowMaterializer()

      val server = new TestServer()
      val client = new TestClient(server.host, server.port)

      val request = RequestBuilding.Get(s"http://${server.host}:${server.port}")

      import akka.http.impl.util._

      // We make an initial request, and stream in the response. This is to ensure all flows are initialized.
      client.sendRequest(request)
        .flatMap(_.entity.toStrict(defaultTimeout))
        .map(_.data).awaitResult(defaultTimeout) shouldEqual server.responseBody

      val initialActorCount: Int = Tools.countActorsInSystem(server.actorSystem).values.sum

      // We make a request, and shut down the client's actor system immediately after receiving the metadata -
      // that is to say, before the server has had time to dispatch the response body.
      //
      // This will cause the server to leak one [[ActorInterpreter]]
      client.sendRequest(request)
        .andThen {
          case Success(response) ⇒
            client.actorSystem.shutdown()
            Thread.sleep(server.responseLatency.toMillis * 2)
        }.awaitResult(defaultTimeout) // shouldEqual ()

      // Then
      /**
       * We wait to allow all streams enough time to complete. Under a version where the bug doesn't exist,
       * 1 second is enough for the [[ActorInterpreter]] count to fall back to its initial value.
       */
      val cooldownAllowance = 1.seconds
      Thread.sleep(cooldownAllowance.toMillis)

      val finalActorCount = Tools.countActorsInSystem(server.actorSystem).values.sum

      println(initialActorCount, finalActorCount)
      /**
       * In the latest version, this will fail - the final count will be precisely 1 greater than the initial count.
       */
      finalActorCount should be <= initialActorCount

      server.actorSystem.shutdown()
      testSystem.shutdown()
    }
  }
}
```

Here's my alternative `countActorsInSystem` method that prints out information about what kind of interpreters have leaked. It would also help if you run your original test with this method and post the output:

```scala
object Tools {
  /**
   * A method that counts the actors of each type inside a given actor system.
   */
  def countActorsInSystem(system: ActorSystem) = {

    val actorCountsByClass = new mutable.Map.WithDefault[Class[_ <: Actor], Int](mutable.Map.empty, _ ⇒ 0)

    def walk(ref: ActorRef): Unit = {
      ref match {
        case c: ActorRefWithCell ⇒
          c.underlying match {
            case ac: ActorCell ⇒
              if (ac.actor != null) {
                val actorClass: Class[_ <: Actor] = ac.actor.getClass
                if (actorClass == classOf[ActorInterpreter]) {
                  ac.actor.asInstanceOf[ActorInterpreter].ops.foreach { stage ⇒
                    println(s"${stage.getClass.getSimpleName} in state" +
                      s" ${Option(ac.actor).map(_.receive.getClass.getSimpleName).getOrElse("")}")
                  }

                }
                actorCountsByClass.put(actorClass, actorCountsByClass(actorClass) + 1)
              }
              c.children.foreach(walk)

            case unrecognizedCell: Cell ⇒
              unrecognizedCell.childrenRefs.children.foreach(walk)
          }
        case _ ⇒
      }
    }

    walk(system.asInstanceOf[ActorSystemImpl].lookupRoot)

    actorCountsByClass
  }
}
```

**Comment by drewhk** [comment]
hm, we should put this into the stream testkit, and probably integrate with assertAllStagesStopped.

**Comment by jrudolph** [comment]
Btw, the "in state ..." line doesn't ever print anything useful because it always prints `OrElse` which comes from composing the `Receive`s inside of the ActorInterpreter implementation (also it's on the wrong level inside the `foreach`).

@drewhk you would probably know what kind of useful extra information could be shown from ActorInterpreter instead.

**Comment by jrudolph** [comment]
@catalin-ursachi  I (somewhat accidentally) removed your usage of `akka.pattern.after` because the import got lost when I copied the code into IntelliJ and then I didn't see where it came from. As I said, I did only a very superficial testing...

**Comment by catalin-ursachi** [comment]
(As a side question: is there any SBT command for running a single Akka test?)

**Comment by jrudolph** [comment]
Yep, `test-only akka.http.IncomingConnectionActorInterpreterLeakSpec2 -- -z "<part of description text>"`.

**Comment by jrudolph** [comment]
The `-- -z "..."` part is only necessary if you want to single out tests within one class (in case that wasn't obvious).

**Comment by drewhk** [comment]
I will also run the test, but currently I am a bit occupied.

**Comment by catalin-ursachi** [comment]
Thanks!

Confirmed that your test passes for me on both release-2.3-dev and 0af8c42, so long as I don't upgrade push the Akka version up to 2.3.11.

**Comment by jrudolph** [comment]
So, does that mean that the test is broken or something is wrong when using Akka 2.3.11?

**Comment by drewhk** [comment]
Looking at the milestone:
https://github.com/akka/akka/issues?q=milestone%3A2.3.11

the suspects can be:
 - actor starvation fix. This is unlikely however, since this is a revert of a change to a state that worked in many versions in a row for a long time
 - BatchingExecutor, but I am not sure if it is related
 - DirectByteBuffer pool fix. This is the most likely offender

**Comment by catalin-ursachi** [comment]
Okay, never mind the above: the test *fails* on release-2.3-dev and older versions, *independently* of whether we use 2.3.10 or 2.3.11.

The test code should read:
```scala
def slowSource() = {

      val p = Promise[ByteString]()
      actorSystem.scheduler.scheduleOnce(responseLatency) {
        p.complete(Success(responseBody))
      }
      Source(p.future)
    }
```
in place of
```scala
val p = Promise[ByteString]()
    actorSystem.scheduler.scheduleOnce(responseLatency) {
      p.complete(Success(responseBody))
    }

    def slowSource() = Source(p.future)
```

Because otherwise, by the time we get to the request we're trying to break, we'll have already made a request and waited for the response, so the future will already have completed. So the test would fail only if the RST was sent in the precise instant before the response body could be dispatched.

With the substitution described above, the test (rightly) fails on all versions.

**Comment by jrudolph** [comment]
Nice and I apologize for the sloppy test translation before.

**Comment by drewhk** [comment]
*phew* that sounds less scary now

**Comment by jrudolph** [comment]
Yes, that seems to be still the Websocket integration. It seems there's no `CompletionHandling` for a `FlexiMerge` when the downstream is finished, so I'm not sure how to deal with that properly.

**Comment by drewhk** [comment]
Ah, yes, cancellation. That needs to be solved.

**Comment by drewhk** [comment]
preStart/postStop support for FlexiX should be ok for this case, right?

**Comment by jrudolph** [comment]
Hmm, it needs access to the state in some way but if postStop could access a `var` of the instance it would probably be enough.

**Comment by jrudolph** [comment]
Fix in #17728 that uses the new postStop introduced in #17638.
